zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,292 @@
+#![allow(dead_code, unused_variables)]
+use super::{Transaction, TransparentInput};
+use crate::{
+    parameters::ConsensusBranchId, serialization::ZcashSerialize, types::BlockHeight, Network,
+    NetworkUpgrade,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use std::io;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+pub const SIGHASH_ALL: u32 = 1;
+const SIGHASH_NONE: u32 = 2;
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub struct SigHasher<'a> {
+    pub(crate) trans: &'a Transaction,
+    pub(crate) hash_type: u32,
+    pub(crate) network: Network,
+    pub(crate) height: BlockHeight,
+}
```

Comment:
The `SigHasher` type has no purpose other than to organize the sighash code -- its data is exactly the parameters to the `Transaction::sighash` function defined above. So I don't think we should create it.

Reply:
Its purpose is to organize all the testable subcomponents of the sighash algo into methods and to share state between the various methods via `&self` rather than duplicating args in a ton of functions making everything way noiser to read. I've reduced the scope of the export because this shouldn't be `pub` but I'd prefer to keep it as is.